Source/DafnyCore/Compilers/SinglePassCompiler.cs

```diff
@@ -1047,9 +1049,134 @@ public class TypeArgumentInstantiation {
+        }
+      }
+
+      return MemberCompileStatus.Ordinary;
     }
+
+    protected bool IsInvisibleWrapper(DatatypeDecl dt, out DatatypeDestructor coreDestructor) {
+      if (!DafnyOptions.O.OptimizeInvisibleDatatypeWrappers ||
+          !OptimizesInvisibleDatatypeWrappers ||
+          !(dt is IndDatatypeDecl) ||
+          dt.IsExtern(out _, out _) ||
+          dt.Members.Any(member => member is Field)) {
+        // don't optimize
+      } else {
+        var i = dt.Ctors.FindIndex(ctor => !ctor.IsGhost);
+        if (0 <= i) {
+          if (dt.Ctors.FindIndex(i + 1, ctor => !ctor.IsGhost) == -1) {
+            // there is exactly one non-ghost constructor
+            var ctor = dt.Ctors[i];
+            var j = ctor.Destructors.FindIndex(dtor => !dtor.IsGhost);
+            if (0 <= j) {
+              if (ctor.Destructors.FindIndex(j + 1, dtor => !dtor.IsGhost) == -1) {
+                // there is exactly one non-ghost parameter to "ctor"
+                var candidateCoreDestructor = ctor.Destructors[j];
+                var properlyContainedTypeDecls = new HashSet<TopLevelDecl>();
+                CompiledTypes(candidateCoreDestructor.Type, properlyContainedTypeDecls);
+                if (!properlyContainedTypeDecls.Contains(dt)) {
+                  coreDestructor = candidateCoreDestructor;
+                  return true;
+                }
+              }
+            }
+          }
+        }
+      }
+      coreDestructor = null;
+      return false;
+    }
+
+    /// <summary>
+    /// This method is for types the analogous thing to "FreeVariables(e)" for an expression "e".
+    /// It returns the user-defined types that are part of "type" (expanding past subset types).
+    /// </summary>
+    void CompiledTypes(Type type, ISet<TopLevelDecl> typeDecls) {
+      type = type.NormalizeExpand();
+      var decl = (type as UserDefinedType)?.ResolvedClass;
+      if (decl != null && typeDecls.Add(decl) && decl is DatatypeDecl dt) {
+        // we just added "decl" to the set
+        foreach (var ctor in dt.Ctors.Where(ctor => !ctor.IsGhost)) {
+          foreach (var dtor in ctor.Destructors.Where(dtor => !dtor.IsGhost)) {
+            CompiledTypes(dtor.Type, typeDecls);
+          }
+        }
+      }
+      type.TypeArgs.ForEach(ty => CompiledTypes(ty, typeDecls));
+    }
+
+    protected bool CanBeLeftUninitialized(Type type) {
+      if (type.NormalizeExpandKeepConstraints() is UserDefinedType udt && udt.ResolvedClass is DatatypeDecl dt) {
+        if (dt.GetGroundingCtor().IsGhost) {
+          return true;
+        }
+        if (IsInvisibleWrapper(dt, out var dtor)) {
+          var typeSubst = TypeUtil.TypeSubstitutionMap(dt.TypeArgs, udt.TypeArgs);
+          return CanBeLeftUninitialized(TypeUtil.SubstType(dtor.Type, typeSubst));
+        }
+      }
+      return false;
+    }
+
+    /// <summary>
+    /// Remove any invisible type wrappers and simplify ghost typle types.
```

Comment:
I'm quite looking forward to seeing this be implemented as a separate pass. I think it'll improve the readability a lot.